lib/linked_list.rb

```diff
@@ -18,95 +18,195 @@ def initialize
+      current_node = @head
+      
+      until current_node.nil?
+        return true if current_node.data == value
+        current_node = current_node.next
+      end
+      
+      return false
     end
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max
-      raise NotImplementedError
+      return nil if @head.nil?
+
+      current_node = @head
+      max = 0
+      until current_node.nil?
+        if current_node.data > max
+          max = current_node.data
+        end
+        current_node = current_node.next
+      end
+
+      return max
     end
 
     # method to return the min value in the linked list
     # returns the data value and not the node
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_min
-      raise NotImplementedError
+      return nil if @head.nil?
+
+      current_node = @head
+      min = current_node.data
+
+      until current_node.nil?
+        if current_node.data < min
+          min = current_node.data
+        end
+        current_node = current_node.next          
+      end
+
+      return min
     end
 
 
     # Additional Exercises 
     # returns the value in the first node
     # returns nil if the list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def get_first
-      raise NotImplementedError
+      return nil if @head.nil?   # I don't understand why this is not passing the test
+
+      return @head.data
     end
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def add_last(value)
-      raise NotImplementedError
+      new_node = Node.new(value)
+
+      current_node = @head
+      if current_node.nil?
+        return @head = new_node
+      end
+      
+      until current_node.next.nil?
+        current_node = current_node.next
+      end
+
+      current_node.next = new_node
     end
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def length
-      raise NotImplementedError
+      current_node = @head
+      counter = 0
+      until current_node.nil?
+        counter += 1
+        current_node = current_node.next
+      end
+
+      return counter
     end
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns nil if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(index)
-      raise NotImplementedError
+      length = self.length
+
+      return nil if index > length

```

Comment:
This would fix the failing test

```suggestion
      return nil if index >= length

```